Client/WebFilters/FilterListResourceDownloader.swift

```diff
@@ -0,0 +1,483 @@
+          isEnabled: isEnabled,
+          componentId: allFilterListSettings[index].componentId
+        )
+      } else {
+        create(
+          uuid: uuid,
+          componentId: nil,
+          isEnabled: isEnabled
+        )
+      }
+    }
+    
+    /// Set the enabled status and componentId of a filter list setting if the setting exists.
+    /// Otherwise it will create a new setting with the specified properties
+    ///
+    /// - Warning: Do not call this before we load core data
+    public func upsertSetting(uuid: String, isEnabled: Bool, componentId: String?) {
+      if allFilterListSettings.contains(where: { $0.uuid == uuid }) {
+        updateSetting(
+          uuid: uuid,
+          isEnabled: isEnabled,
+          componentId: componentId
+        )
+      } else {
+        create(
+          uuid: uuid,
+          componentId: componentId,
+          isEnabled: isEnabled
+        )
+      }
+    }
+    
+    /// Set the enabled status of a filter list setting
+    ///
+    /// - Warning: Do not call this before we load core data
+    public func set(folderURL: URL, forUUID uuid: String) {
+      guard let index = allFilterListSettings.firstIndex(where: { $0.uuid == uuid }) else {
+        return
+      }
+      
+      guard allFilterListSettings[index].folderURL != folderURL else { return }
+      allFilterListSettings[index].folderURL = folderURL
+      FilterListSetting.save(inMemory: inMemory)
+    }
+    
+    private func updateSetting(uuid: String, isEnabled: Bool, componentId: String?) {
+      guard let index = allFilterListSettings.firstIndex(where: { $0.uuid == uuid }) else {
+        return
+      }
+      
+      guard allFilterListSettings[index].isEnabled != isEnabled || allFilterListSettings[index].componentId != componentId else {
+        // Ensure we stop if this is already in sync in order to avoid an event loop
+        // And things hanging for too long.
+        // This happens because we care about UI changes but not when our downloads finish
+        return
+      }
+        
+      allFilterListSettings[index].isEnabled = isEnabled
+      allFilterListSettings[index].componentId = componentId
+      FilterListSetting.save(inMemory: inMemory)
+    }
+    
+    /// Create a filter list setting for the given UUID and enabled status
+    private func create(uuid: String, componentId: String?, isEnabled: Bool) {
+      let setting = FilterListSetting.create(uuid: uuid, componentId: componentId, isEnabled: isEnabled, inMemory: inMemory)
+      allFilterListSettings.append(setting)
+    }
+  }
+  /// A shared instance of this class
+  ///
+  /// - Warning: You need to wait for `DataController.shared.initializeOnce()` to be called before using this instance
+  public static let shared = FilterListResourceDownloader()
+  
+  /// Object responsible for getting component updates
+  private var adBlockService: AdblockService?
+  /// Manager that handles updates to filter list settings in core data
```

Comment:
Not sure if this is GitHub formatting issue… but it looks like every line after is not indented?